tinygrad/lazy.py

```diff
@@ -209,11 +209,32 @@ def shuffle_and_prune_movement_ops(self, st: ShapeTracker, op: MovementOps, arg:
         return root.reshape(ret.st.shape)
     return ret
 
-  def reduce_op(self:LazyBuffer, op:ReduceOps, new_shape:Tuple[int, ...]) -> LazyBuffer:
+  def _reduce_op(self:LazyBuffer, op:ReduceOps, new_shape:Tuple[int, ...]) -> LazyBuffer:
     if self.shape == tuple(new_shape): return self
     srcs = _push_movement_ops((self,)) if SHUFFLE_MOVEMENT_OPS else (self,)
     return create_lazybuffer(self.device, ShapeTracker(new_shape), ReduceOps, LazyOp(op, srcs, new_shape), self.dtype)
 
+  def reduce_op(self:LazyBuffer, op:ReduceOps, new_shape:Tuple[int, ...]) -> LazyBuffer:
+    if prod(self.shape) // prod(new_shape) > 16384:
+      reduced_dimensions = [(i, divisor := math.gcd(512, old), divisor/(stride or math.inf)) for i, (old, new, stride) in enumerate(zip(self.shape, new_shape, self.st.real_strides())) if old != new] # type: ignore
+      other_dimensions = [old for (old, new) in zip(self.shape, new_shape) if old == new]
+
+      # heuristic -> choose largest divisor to split on, penalize large strides
+      dimension_to_split, divisor, heuristic = max(reduced_dimensions, key=lambda v: v[2])
+      if heuristic < 1: return self._reduce_op(op, new_shape)
+
+      # When global/local dims is filled well, no need to split making it overloaded.
+      sm_count = 32
+      if Device[self.device].device_info: sm_count = unwrap_optional(Device[self.device].device_info.cores_count_executing_in_parallel, sm_count) # type: ignore
```

Comment:
Can a generic estimate / static value be used for this? Then `DeviceInfo` doesn't have to be in this PR.